### ISSUE TITLE
Arr_trapezoid_ric_point_location crashes on edge removal

### ISSUE THREAD

**Issue by chgans** [CGAL/cgal#1955]
## Issue Details

Arr_trapezoid_ric_point_location crashes when removing an half-edge that is connected to other half-edges.

## Source Code

```c++
#include <CGAL/Cartesian.h>
#include <CGAL/Arr_segment_traits_2.h>
#include <CGAL/Arrangement_2.h>
#include <CGAL/Arr_walk_along_line_point_location.h>
#include <CGAL/Arr_trapezoid_ric_point_location.h>

typedef CGAL::Exact_predicates_exact_constructions_kernel Kernel;
typedef Kernel::FT                                        Number_type;
typedef CGAL::Arr_segment_traits_2<Kernel>                Traits;
typedef Traits::Point_2                                   Point;
typedef Traits::X_monotone_curve_2                        Segment;
typedef Traits::Curve_2                                   Curve;
typedef CGAL::Arrangement_2<Traits>                       Arrangement;

typedef CGAL::Arr_trapezoid_ric_point_location<Arrangement>   TrapezoidLocator;

int main (int /*argv**/, char ** /*argv*/)
{
    Arrangement arrangement;
    TrapezoidLocator locator(arrangement); 

    auto v1 = arrangement.insert_in_face_interior(Point(0, 0), arrangement.unbounded_face());
    auto v2 = arrangement.insert_in_face_interior(Point(0, 1), arrangement.unbounded_face());
    auto v3 = arrangement.insert_in_face_interior(Point(1, 1), arrangement.unbounded_face());
    auto v4 = arrangement.insert_in_face_interior(Point(1, 0), arrangement.unbounded_face());
    auto e1 = arrangement.insert_at_vertices(Segment(v1->point(), v2->point()), v1, v2);
    auto e3 = arrangement.insert_at_vertices(Segment(v3->point(), v4->point()), v3, v4);

    // This is fine:
    arrangement.remove_edge(e1, false, false);
    arrangement.remove_edge(e3, false, false);


    // But this will error out when calling remove_edge()
    // what():  CGAL ERROR: assertion violation!
    // Expr: he == first
    // File: /usr/include/CGAL/Arr_point_location/Trapezoidal_decomposition_2_impl.h
    // Line: 1542
    e1 = arrangement.insert_at_vertices(Segment(v1->point(), v2->point()), v1, v2);
    e3 = arrangement.insert_at_vertices(Segment(v3->point(), v4->point()), v3, v4);    
    auto e2 = arrangement.insert_at_vertices(Segment(v2->point(), v3->point()), v2, v3);
    arrangement.remove_edge(e2, false, false); // ERROR!!!

    return 0;
}
```

```sh
g++ -o test main.cpp -lCGAL -lgmp -lmpfr
./test
terminate called after throwing an instance of 'CGAL::Assertion_exception'
  what():  CGAL ERROR: assertion violation!
Expr: he == first
File: /usr/include/CGAL/Arr_point_location/Trapezoidal_decomposition_2_impl.h
Line: 1542
Aborted (core dumped)
```
## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Linux KUbuntu 16.10, 64 bits
* Compiler: pool/main/g/gcc-defaults/gcc_6.1.1-1ubuntu2_amd64.deb
* Release or debug mode: Debug mode
* Specific flags used (if any): None
* CGAL version: pool/universe/c/cgal/libcgal-dev_4.8.1-1build1_amd64.deb
* Boost version: pool/main/b/boost-defaults/libboost-dev_1.61.0.2_amd64.deb
* Other libraries versions if used (Eigen, TBB, etc.): None

**Comment by efifogel** [comment]
Apparently, the assertion statement that causes the violation should simply
be removed (and a couple of redundant statement that precede the assertion).
Anyhow, the attached patch fixes it.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by chgans** [comment]
Hi Efi,

Thanks for looking at this. I couldn't find your patch anywhere, does it simply remove the assert? I would be interested to see it in case you have removed more assert/warning.

Here is another similar, valid and simple case:
This will trigger a warning with a reference to bug reporting, I don't known if it's harmless or not.

```C++

#include <CGAL/Cartesian.h>
#include <CGAL/Arr_segment_traits_2.h>
#include <CGAL/Arrangement_2.h>
#include <CGAL/Arr_walk_along_line_point_location.h>
#include <CGAL/Arr_trapezoid_ric_point_location.h>

typedef CGAL::Exact_predicates_exact_constructions_kernel Kernel;
typedef Kernel::FT                                        Number_type;
typedef CGAL::Arr_segment_traits_2<Kernel>                Traits;
typedef Traits::Point_2                                   Point;
typedef Traits::X_monotone_curve_2                        Segment;
typedef Traits::Curve_2                                   Curve;
typedef CGAL::Arrangement_2<Traits>                       Arrangement;

typedef CGAL::Arr_trapezoid_ric_point_location<Arrangement>   TrapezoidLocator;

int main (int /*argv**/, char ** /*argv*/)
{
    Arrangement arrangement;
    TrapezoidLocator locator(arrangement);

    // First create a new face
    auto v1 = arrangement.insert_in_face_interior(Point(0, 0), arrangement.unbounded_face());
    auto v2 = arrangement.insert_in_face_interior(Point(0, 1), arrangement.unbounded_face());
    auto v3 = arrangement.insert_in_face_interior(Point(1, 1), arrangement.unbounded_face());
    auto v4 = arrangement.insert_in_face_interior(Point(1, 0), arrangement.unbounded_face());
    auto e1 = arrangement.insert_at_vertices(Segment(v1->point(), v2->point()), v1, v2);
    auto e2 = arrangement.insert_at_vertices(Segment(v2->point(), v3->point()), v2, v3);
    auto e3 = arrangement.insert_at_vertices(Segment(v3->point(), v4->point()), v3, v4);
    auto e4 = arrangement.insert_at_vertices(Segment(v4->point(), v1->point()), v4, v1);

    // To trigger the warning, we need to add an extra inner CCB in the unbounded face ...
    auto v5 = arrangement.insert_in_face_interior(Point(2, 2), arrangement.unbounded_face());
    auto v6 = arrangement.insert_in_face_interior(Point(3, 3), arrangement.unbounded_face());
    auto e5 = arrangement.insert_at_vertices(Segment(v5->point(), v6->point()), v5, v6);

    // ... then start removing the edges
    arrangement.remove_edge(e5);
    arrangement.remove_edge(e4);
    arrangement.remove_edge(e3);
    arrangement.remove_edge(e2);
    arrangement.remove_edge(e1); // This one will generate a warning

    return 0;
}
```
Which gives:
```
CGAL warning: check violation!
Expression : lt1==POINT && lt2==POINT
File       : /usr/include/CGAL/Arr_point_location/Trapezoidal_decomposition_2_impl.h
Line       : 1781
Explanation: 
Refer to the bug-reporting instructions at http://www.cgal.org/bug_report.html
```

**Comment by efifogel** [comment]
Hi Christian,

1. The patch is simple: just remove the lines below near 1542.

-  if (he->source() == v)
-    he = he->twin();
-  CGAL_assertion(he == first);

Naturally, we'll fix it in our sources.

2. The new test case reveals another problem. (Or, at least it seems so.)
I would appreciate it if you submit a new issue.

Thanks,
Efi

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by chgans** [comment]
Thanks Efi, I've created a separate issue: #1967

**Comment by sloriot** [comment]
Fixed by #1962
